### PR TITLE
Derived types codegen (part 1)

### PR DESCRIPTION
**Pull request by schweitzpgi**
Get started on cleaning up the IR, etc.
  Code gen for alloca.
  Add initial test.

### CODE REVIEW

**Review by jeanPerier — COMMENTED** [review]
Overall looks good to me to deal with PDTs in a layout agnostic way.
It's however not completely clear to me how f18 PDT design with descriptors for automatic components plugs into it.
For instance, I am not quite sure how an inlined allocation would work. It looks like to me lowering would have to do make the "hidden" descriptors visible so that the automatic data can actually be allocated and mapped back in the descriptor field directly in lowering. On the other hand, if we plan on using f18 runtime to allocate such derived types then that may be OK.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 1]
```diff
@@ -325,13 +338,41 @@ struct AllocaOpConversion : public FIROpConversion<fir::AllocaOp> {
...
-    rewriter.replaceOpWithNewOp<mlir::LLVM::AllocaOp>(alloc, ty, size,
+    auto resultTy = ty;
+    if (alloc.hasLenParams()) {
+      unsigned end = alloc.numLenParams();
+      llvm::SmallVector<mlir::Value, 4> lenParams;
+      for (; i < end; ++i)
+        lenParams.push_back(operands[i]);
+      auto i64Ty = mlir::IntegerType::get(alloc.getContext(), 64);
+      auto recTy = alloc.getInType().dyn_cast<fir::RecordType>();
+      assert(recTy && "expected fir.type");
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:352]: Right. We should probably make fir.char act more like a dependent type and less like an array (as it used to).

File: flang/test/Fir/pdt.fir [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+func private @bar(!fir.ref<!fir.char<1,?>>)
+
+// CHECK-LINE: define i8* @_QPfoo(i32 %0, i32 %1)
+func @_QPfoo(%arg0 : i32, %arg1 : i32) {
+  // CHECK: %[[size:.*]] = call i64 @_QTt1P.mem.size(i32 %0, i32 %1)
+  // CHECK: %[[alloc:.*]] = alloca i8, i64 %[[size]]
+  %0 = fir.alloca !fir.type<_QTt1(p1:i32,p2:i32){f1:!fir.char<1,?>,f2:!fir.char<1,?>}>(%arg0, %arg1 : i32, i32)
+  %1 = fir.field_index f2, !fir.type<_QTt1>(%arg0, %arg1 : i32, i32)
+  //%2 = fir.coordinate_of %0, %1 : (!fir.ref<!fir.type<_QTt1>>, !fir.field) -> !fir.ref<!fir.char<1,?>>
+  %2 = fir.zero_bits !fir.ref<!fir.char<1,?>>
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/test/Fir/pdt.fir:108]: I assume you're referring to the runtime. This work isn't precluding the runtime, of course, but lowering will have to craft derived types with descriptors to use the runtime.

fir.alloca will (eventually) work as expected in either case.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 3]
```diff
@@ -325,13 +338,41 @@ struct AllocaOpConversion : public FIROpConversion<fir::AllocaOp> {
...
+    unsigned i = 0;
+    auto size = genConstantIndex(loc, ity, rewriter, 1).getResult();
     auto ty = convertType(alloc.getType());
-    rewriter.replaceOpWithNewOp<mlir::LLVM::AllocaOp>(alloc, ty, size,
+    auto resultTy = ty;
+    if (alloc.hasLenParams()) {
+      unsigned end = alloc.numLenParams();
+      llvm::SmallVector<mlir::Value, 4> lenParams;
+      for (; i < end; ++i)
+        lenParams.push_back(operands[i]);
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:349]: Right. They are different. The arguments to the Op being converted are unreliable. They may be converted or not. If we assume anything about them, we get burned. The arguments passed in `operands` are reliable and must have been converted.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 4]
```diff
@@ -315,6 +315,19 @@ struct AddrOfOpConversion : public FIROpConversion<fir::AddrOfOp> {
...
+/// type. The size of the object may depend on the LEN type parameters of the
+/// derived type.
+static mlir::LLVM::LLVMFuncOp
+getDependentTypeMemSizeFn(fir::RecordType recTy, fir::AllocaOp op,
+                          mlir::ConversionPatternRewriter &rewriter) {
+  auto module = op->getParentOfType<mlir::ModuleOp>();
+  std::string name = recTy.getName().str() + "P.mem.size";
+  if (auto memSizeFunc = module.lookupSymbol<mlir::LLVM::LLVMFuncOp>(name))
+    return memSizeFunc;
+  TODO_NOLOC("did not find allocation function");
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:328]: Thanks. Will do.

**Review by jeanPerier — APPROVED** [review]

**Pull request merged**